### PR TITLE
[SYNPY-1419] Httpx async client

### PR DESCRIPTION
**Pull request by BryanFauble**
**Problem:**

1. The default requests library does not contain any async interfaces
2. The back-off was not sufficient and the retry logic needs some TLC

**Solution:**

1. Adding the HTTPX dependency
3. Creating 2 new async client instances in Synapse that can be used to interact with Synapse or the Storage Providers
4. Creating a new API class with async interfaces that work directly with the the [EntityBundleV2Controller](https://rest-docs.synapse.org/rest/#org.sagebionetworks.repo.web.controller.EntityBundleV2Controller)
5. I re-wrote our exponential back-off logic to follow: 1ms + random(10ms-100ms). The 1ms is exponentially backed off by a factor of 2 for each retry and the random is not touched.
6. Updated retry logic to not fail based on the number of retries, instead fail after a certain amount of time has elapsed.

**Testing:**

1. Integration/Unit tests
2. Manual testing with these scripts:

**Creating and then deleting a folder**
```
import asyncio
import uuid

import synapseclient
from synapseclient.api import post_entity_bundle2_create


syn = synapseclient.Synapse(debug=True)
syn.login()


async def execute_create_then_delete_folder():
    folder_entity = {
        "entity": {
            "concreteType": "org.sagebionetworks.repo.model.Folder",
            "name": "test_folder_" + str(uuid.uuid4()),
            "parentId": "syn53972896",
        }
    }
    result = await post_entity_bundle2_create(request=folder_entity)
    print(result)

    await syn.rest_delete_async(uri=f"/entity/{result['entity']['id']}")
    await syn.rest_get_async(uri=f"/entity/{result['entity']['id']}")


asyncio.run(execute_create_then_delete_folder())
```

**Creating a folder**
```
import asyncio
import uuid

import synapseclient
from synapseclient.api import post_entity_bundle2_create


syn = synapseclient.Synapse(debug=True)
syn.login()


async def execute_create_folder():
    folder_entity = {
        "entity": {
            "concreteType": "org.sagebionetworks.repo.model.Folder",
            "name": "test_folder_" + str(uuid.uuid4()),
            "parentId": "syn53972896",
        }
    }
    result = await post_entity_bundle2_create(request=folder_entity)
    print(result)


asyncio.run(execute_create_folder())
```

**Getting a specific version of a file**
```
import asyncio

import synapseclient
from synapseclient.api import get_entity_id_version_bundle2


syn = synapseclient.Synapse(debug=True)
syn.login()


async def execute_get_with_id_and_version():
    synapse_id = "syn53604544"
    result = await get_entity_id_version_bundle2(entity_id=synapse_id, version=2)
    print(result)


asyncio.run(execute_get_with_id_and_version())
```

**Updating an entity by the /bundle2 endpoints**
```
import asyncio
import uuid

import synapseclient
from synapseclient.api import get_entity_id_bundle2, put_entity_id_bundle2
from synapseclient.annotations import _convert_to_annotations_list


syn = synapseclient.Synapse(debug=True)
syn.login()


async def execute_put():
    synapse_id = "syn53666147"
    result = await get_entity_id_bundle2(entity_id=synapse_id)
    print(result)

    print("\n\n\n\n\n")
    result["entity"]["description"] = str(uuid.uuid4())
    result["entity"]["versionNumber"] = result["entity"]["versionNumber"] + 1
    result["entity"]["versionLabel"] = str(result["entity"]["versionNumber"])
    synapse_annotations = _convert_to_annotations_list(
        {**result["annotations"]["annotations"], "asdf": [str(uuid.uuid4())]}
    )

    result["annotations"]["annotations"] = synapse_annotations

    put_result = await put_entity_id_bundle2(entity_id=synapse_id, request=result)

    print(put_result)


asyncio.run(execute_put())
```

I verified in several cases that the backoff was working as expected via console logs. I also verified several exceptions were being logged to the console based on the error:

HTTP 429:
```
2024-03-07 09:33:09,610 [retry:342 - DEBUG]: retrying on status code: 429 - repo-prod.prod.sagebase.org/repo/v1/entity/syn53666136/bundle2 - Too many concurrent requests. Allowed 3 concurrent connections at any time
```

HTTP 404:
```
Traceback (most recent call last):
  File "/home/bfauble/.pyenv/versions/3.8.18/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/bfauble/.pyenv/versions/3.8.18/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/bfauble/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/home/bfauble/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/home/bfauble/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 284, in run_file
    runpy.run_path(target, run_name="__main__")
  File "/home/bfauble/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 321, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/home/bfauble/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 135, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/home/bfauble/.vscode-server/extensions/ms-python.python-2024.2.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/junk/execute_get_bundle_async.py", line 62, in <module>
    asyncio.run(execute_gets())
  File "/home/bfauble/.pyenv/versions/3.8.18/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/home/bfauble/.pyenv/versions/3.8.18/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/junk/execute_get_bundle_async.py", line 43, in execute_gets
    result = await get_entity_id_bundle2(
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/api/entity_bundle_services_v2.py", line 29, in get_entity_id_bundle2
    return await client.rest_post_async(
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/client.py", line 6211, in rest_post_async
    response = await self._rest_call_async(
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/client.py", line 6143, in _rest_call_async
    self._handle_httpx_synapse_http_error(response)
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/client.py", line 5834, in _handle_httpx_synapse_http_error
    exceptions._raise_for_status_httpx(response, verbose=self.debug)
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/core/exceptions.py", line 277, in _raise_for_status_httpx
    raise SynapseHTTPError(message, response=response)
synapseclient.core.exceptions.SynapseHTTPError: 404 Client Error: Resource: 'syn536661362' does not exist
Resource: 'syn536661362' does not exist

>>>>>> Request <<<<<<
https://repo-prod.prod.sagebase.org/repo/v1/entity/syn536661362/bundle2 POST
>>> Headers: Headers({'host': 'repo-prod.prod.sagebase.org', 'accept-encoding': 'gzip, deflate', 'connection': 'keep-alive', 'content-type': 'application/json; charset=UTF-8', 'accept': 'application/json; charset=UTF-8', 'user-agent': 'synapseclient/4.1.1 python-requests/2.31.0', 'content-length': '118', 'authorization': '[secure]'})
>>> Body: b'{"includeEntity": true, "includeAnnotations": true, "includeFileHandles": true, "includeRestrictionInformation": true}'

>>>>>> Response <<<<<<
<Response [404 ]>
>>> Headers: Headers({'date': 'Thu, 07 Mar 2024 16:51:27 GMT', 'content-type': 'application/json;charset=UTF-8', 'content-length': '122', 'connection': 'keep-alive', 'server': '34 f6e676 27, 1647 57 c616 4796f6 e6 3712 0295. f6 5702 66f657 e6 4602 f657 27 02 3756, 3627 5647e2.', 'x-xss-protection': '1;mode=block', 'x-content-type-options': 'nosniff', 'access-control-allow-origin': '*', 'strict-transport-security': 'max-age=31536000', 'set-cookie': 'sessionID=367181e0-07c4-40c0-be70-04079e227e77; HttpOnly', 'vary': 'Accept-Encoding,User-Agent', 'content-encoding': 'gzip', 'x-frame-options': 'SAMEORIGIN'})
>>> Body: {"concreteType":"org.sagebionetworks.repo.model.ErrorResponse","reason":"Resource: 'syn536661362' does not exist"}
```

HTTP 404 in trash:
```
2024-03-07 11:45:38,701 [client:6189 - ERROR]: Error in rest_get_async
Traceback (most recent call last):
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/client.py", line 6177, in rest_get_async
    response = await self._rest_call_async(
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/client.py", line 6145, in _rest_call_async
    self._handle_httpx_synapse_http_error(response)
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/client.py", line 5833, in _handle_httpx_synapse_http_error
    exceptions._raise_for_status_httpx(response, verbose=self.debug)
  File "/home/bfauble/BryansGreatWorkspace/synapsePythonClient/synapseclient/core/exceptions.py", line 277, in _raise_for_status_httpx
    raise SynapseHTTPError(message, response=response)
synapseclient.core.exceptions.SynapseHTTPError: 404 Client Error: Entity syn53972940 is in trash can.
Entity syn53972940 is in trash can.

>>>>>> Request <<<<<<
https://repo-prod.prod.sagebase.org/repo/v1/entity/syn53972940 GET
>>> Headers: Headers({'host': 'repo-prod.prod.sagebase.org', 'accept-encoding': 'gzip, deflate', 'connection': 'keep-alive', 'content-type': 'application/json; charset=UTF-8', 'accept': 'application/json; charset=UTF-8', 'user-agent': 'synapseclient/4.1.1 python-requests/2.31.0', 'cookie': 'sessionID=0af536fc-19b2-4257-a535-2f848f46cfc0', 'authorization': '[secure]'})
>>> Body: b''

>>>>>> Response <<<<<<
<Response [404 ]>
>>> Headers: Headers({'date': 'Thu, 07 Mar 2024 18:45:33 GMT', 'content-type': 'application/json;charset=UTF-8', 'transfer-encoding': 'chunked', 'connection': 'keep-alive', 'server': '34 f6e676 27, 1647 57 c616 4796f6 e6 3712 0295. f6 5702 66f657 e6 4602 f657 27 02 3756, 3627 5647e2.', 'x-xss-protection': '1;mode=block', 'x-content-type-options': 'nosniff', 'access-control-allow-origin': '*', 'strict-transport-security': 'max-age=31536000', 'vary': 'Accept-Encoding,User-Agent', 'content-encoding': 'gzip', 'x-frame-options': 'SAMEORIGIN'})
>>> Body: {"concreteType":"org.sagebionetworks.repo.model.ErrorResponse","reason":"Entity syn53972940 is in trash can."}
```

### CODE REVIEW

File: synapseclient/api/entity_bundle_services_v2.py
Comment (BryanFauble) [inline synapseclient/api/entity_bundle_services_v2.py:1]: Not planning to spend a lot of time on scripts that will exist within this `api` module - The intention is that these will be replaced by a low-level client generated via the OpenAPI spec

Comment (BryanFauble) [reply synapseclient/api/entity_bundle_services_v2.py:1]: This is also the start of an initial effort to split functionality out of the client.py. I started with this because there are some upcoming tickets I am aware of that will use these `/bundle2` endpoints. So I used this during my manual testing to make sure that everything was functioning as I expected.

File: synapseclient/core/exceptions.py
Comment (jaymedina) [inline synapseclient/core/exceptions.py:86]: I'm not sure how I feel about silently returning `None` for calls to the API. Can we add a warning message here saying that an error was received, likely due to a truncated response?

File: synapseclient/client.py
Comment (jaymedina) [inline synapseclient/client.py:293]: Is this a number you've experimented with and found that it was optimal? I'm seeing the default value for this arg is 100 so I wonder why it was brought down to 25

File: synapseclient/core/exceptions.py
Comment (jaymedina) [inline synapseclient/core/exceptions.py:255]: nit: Do we want to remove the if-nesting and just have `if verbose and message is not None`

File: synapseclient/core/retry.py
Comment (jaymedina) [inline synapseclient/core/retry.py:24]: helpful comment :)

File: synapseclient/client.py
Comment (BryanFauble) [reply synapseclient/client.py:293]: Yes. I did some testing around this and found that lower than 25 impacted performance, higher than 25 didn't have much affect on performance - But had higher HTTP 429 rate limit errors coming back from Synapse.

File: synapseclient/core/exceptions.py
Comment (BryanFauble) [reply synapseclient/core/exceptions.py:255]: Not in this case.

`if message is not None` - We'll `raise SynapseHTTPError(message, response=response)`

However,

`if verbose` - We'll append the request/response information to the message.

File: synapseclient/client.py
Comment (BryanFauble) [reply synapseclient/client.py:6217]: I'll include this in a future PR. I started on https://sagebionetworks.jira.com/browse/SYNPY-1420 - Which builds on-top of these changes, and I want to reduce possible conflicts.

https://github.com/Sage-Bionetworks/synapsePythonClient/tree/SYNPY-1420-async-upload-multi-threaded-mix

File: synapseclient/client.py
Comment (jaymedina) [reply synapseclient/client.py:293]: I see - and this number is the total number of connections to the API a user can make in 1 session, correct? My initial worry was that this cap might be too low for use-cases like the Challenge submissions Nextflow workflows (I'm thinking of PEGS & Dynamic Challenge), but each parallel process that uses the python client runs its own session for the most part, so perhaps this wouldn't be a problem if my understanding of what `max_connections` is is correct.

Comment (BryanFauble) [reply synapseclient/client.py:293]: Yes, that's correct. Each machine running this code will have this limit - and it is not shared as there isn't a way to know.

On the other hand, Synapse will kick back HTTP 429 errors if the concurrent connections are too many (Either by a single client, or a single user across many clients). This was the motivation for changing the retry logic to be time-based.

File: synapseclient/core/retry.py
Comment (jaymedina) [inline synapseclient/core/retry.py:281]: nit: What's the advantage of doing this instead of setting default values for these args?

File: synapseclient/core/retry.py
Comment (jaymedina) [inline synapseclient/core/retry.py:343]: nit: Raising*

File: synapseclient/core/retry.py
Comment (BryanFauble) [reply synapseclient/core/retry.py:281]: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: tests/integration/synapseclient/integration_test.py
Comment (jaymedina) [inline tests/integration/synapseclient/integration_test.py:1323]: Does this entity need to be scheduled for cleanup if you delete it in the next step?

File: synapseclient/client.py
Comment (jaymedina) [reply synapseclient/client.py:293]: I see, that makes sense, thanks for explaining this

File: synapseclient/core/retry.py
Comment (jaymedina) [reply synapseclient/core/retry.py:281]: Good to know!

**Review by jaymedina — APPROVED** [review]
Hey @BryanFauble , this looks like it's in good shape and you've answered all my biggest questions/comments. You mentioned there's another feature branch that will be merging into this one, correct? I'll approve of the branch in its current condition and will leave it up to the reviewer of your other branch for the final approval, once everything is merged together.

**Comment by BryanFauble** [comment]
> Hey @BryanFauble , this looks like it's in good shape and you've answered all my biggest questions/comments. You mentioned there's another feature branch that will be merging into this one, correct? I'll approve of the branch in its current condition and will leave it up to the reviewer of your other branch for the final approval, once everything is merged together.

Thanks @jaymedina I'll be tackling any remaining issues in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1077

**Pull request merged**
